File: leetcode/medium/909_snakesAndLadders.py
from typing import List
import json
from collections import deque

class Solution:
    def snakesAndLadders(self, board: List[List[int]]) -> int:
        n = len(board)
        n2 = n**2
        visited = [[-1]*n for i in range(n)]
        visited[-1][0] = 1

        def getNumByLoc(y, x):
            y = n - y - 1
            num = y * n
            num += x + 1 if (y % 2) == 0 else n - x
            return num

        def getLocByNum(num):
            y = num // n
            rem = num % n
            if rem == 0:
                y -= 1
                rem = n
            x = rem - 1 if ((y % 2) == 0) else n - rem
            return (n - y - 1, x)

        # arr = []
        # for i in range(n):
        #     arr.append([])
        #     for j in range(n):
        #         num = getNumByLoc(i, j)
        #         arr[-1].append({
        #             num,
        #             getLocByNum(num),
        #         })

        stepsQueue = deque()
        stepsQueue.append((1, 0))

        while stepsQueue:
            num, steps = stepsQueue.popleft()
            y, x = getLocByNum(num)

            steps += 1

            for val in range(num + 1, min(num + 7, n2+1)):
                if val == n2:
                    return steps

                y, x = getLocByNum(val)
                if visited[y][x] == 1:
                    continue
                visited[y][x] = 1

                if board[y][x] == -1:
                    stepsQueue.append((val, steps))
                    continue

                jump = board[y][x]
                if jump == n2:
                    return steps
                y, x = getLocByNum(jump)
                # Second jump is not visited. I can apply second jump only step by step.
                # if visited[y][x] == 1:
                #     continue
                # visited[y][x] = 1
                stepsQueue.append((jump, steps))


        return -1

'''

[
[-1,-1],
[-1, 3]
]



[
[-1, 1, 2,-1],
[ 2,13,15,-1],
[-1,10,-1,-1],
[-1, 6, 2, 8]
]

[
[-1,-1,-1,46,47,-1,-1,-1],
[51,-1,-1,63,-1,31,21,-1],
[-1,-1,26,-1,-1,38,-1,-1],
[-1,-1,11,-1,14,23,56,57],
[11,-1,-1,-1,49,36,-1,48],
[-1,-1,-1,33,56,-1,57,21],
[-1,-1,-1,-1,-1,-1,2,-1],
[-1,-1,-1,8,3,-1,6,56]
]









'''

def test ():
    params = [
        {
            'input': [[-1,-1,-1,-1,-1,-1],[-1,-1,-1,-1,-1,-1],[-1,-1,-1,-1,-1,-1],[-1,35,-1,-1,13,-1],[-1,-1,-1,-1,-1,-1],[-1,15,-1,-1,-1,-1]],
            'output': 4,
        },
        {
            'input': [[-1,-1],[-1,3]],
            'output': 1,
        },
        {
            'input': [[-1,1,2,-1],[2,13,15,-1],[-1,10,-1,-1],[-1,6,2,8]],
            'output': 2,
        },

        {
            'input': [[-1,-1,-1,46,47,-1,-1,-1],[51,-1,-1,63,-1,31,21,-1],[-1,-1,26,-1,-1,38,-1,-1],[-1,-1,11,-1,14,23,56,57],[11,-1,-1,-1,49,36,-1,48],[-1,-1,-1,33,56,-1,57,21],[-1,-1,-1,-1,-1,-1,2,-1],[-1,-1,-1,8,3,-1,6,56]],
            'output': 4,
        },
        {
            'input': [[-1,-1,27,13,-1,25,-1],[-1,-1,-1,-1,-1,-1,-1],[44,-1,8,-1,-1,2,-1],[-1,30,-1,-1,-1,-1,-1],[3,-1,20,-1,46,6,-1],[-1,-1,-1,-1,-1,-1,29],[-1,29,21,33,-1,-1,-1]],
            'output': 4,
        },


    ]
    solution = Solution()

    for param in params:
        result = solution.snakesAndLadders(param['input'])
        correct = json.dumps(result) == json.dumps(param['output'])

        msg = 'SUCCESS' if json.dumps(result) == json.dumps(param['output']) else 'ERROR'
        msg += '\n'
        if not correct:
            msg += 'input ' + json.dumps(param['input']) + '\n'
            msg += 'output ' + json.dumps(param['output']) + '\n'
            msg += 'result ' + json.dumps(result) + '\n'

        print(msg)


if __name__ == '__main__':
    test()

File: leetcode/medium/909_snakesAndLadders.ts
function snakesAndLadders(board: number[][]): number {
    const visited: number[] = new Array(board.length**2).fill(0);
    let res: number = Number.MAX_VALUE;

    const getLocById = (id: number): number[] => {
        const rowRem: number = Math.ceil(id / board.length);
        const row: number = board.length - rowRem;

        let rem: number = id % board.length;
        let col: number = (board.length + rem - 1) % board.length;
        if ((rowRem % 2) === 0) {
            col = (board.length - rem) % board.length;
        }
        return [row, col];
    }

    const dfs = (id: number, steps: number): void => {
        if (id === board.length**2) {
            res = Math.min(res, steps);
            return;
        }

        for (let i: number = 1; i < 7; ++i) {
            const nei: number = id + 1;
            if (nei > board.length**2 || visited[nei] === 1) {
                continue;
            }
            visited[nei] = 1;
            const [r, c] = getLocById(nei);

            if (board[r][c] === -1) {
                dfs(nei, steps + 1);
            } else {
                dfs(board[r][c], steps + 1);
            }
        }
    }
    // dfs(1, 0);

    // console.log(
    //     'id', 36, getLocById(36),
    //     'id', 35, getLocById(35),
    //     'id', 34, getLocById(34),
    //     'id', 31, getLocById(31),
    //     'id', 25, getLocById(25),
    //     'id', 30, getLocById(30),
    //     'id', 24, getLocById(24),
    //     'id', 19, getLocById(19),
    //     'id', 1, getLocById(1),
    //     'id', 6, getLocById(6),
    // )

    const queue: number[][] = [[1, 0]];

    while (queue.length) {
        const [id, steps] = queue.shift();

        if (id === board.length**2) {
            res = Math.min(res, steps);
            continue;
        }

        for (let i: number = 1; i < 7; ++i) {
            const nei: number = id + i;
            if (nei > board.length**2 || visited[nei] === 1) {
                continue;
            }

            visited[nei] = 1;
            const [r, c] = getLocById(nei);

            if (board[r][c] === -1) {
                queue.push([nei, steps + 1]);
            } else {
                queue.push([board[r][c], steps + 1]);
            }
        }
    }

    return res === Number.MAX_VALUE ? -1 : res;
};

/*

[
[-1,-1,-1,-1,-1,-1],
[-1,-1,-1,-1,-1,-1],
[-1,-1,-1,-1,-1,-1],
[-1,35,-1,-1,13,-1],
[-1,-1,-1,-1,-1,-1],
[-1,15,-1,-1,-1,-1]
]

*/

const test = () => {
    const params = [
        {
            input: [[-1,-1,-1,-1,-1,-1],[-1,-1,-1,-1,-1,-1],[-1,-1,-1,-1,-1,-1],[-1,35,-1,-1,13,-1],[-1,-1,-1,-1,-1,-1],[-1,15,-1,-1,-1,-1]],
            output: 4,
        },
        {
            input: [
                [-1,-1],
                [-1,3]
            ],
            output: 1,
        },
    ];

    params.forEach(({input, output}) => {
        const result = snakesAndLadders(input);

        console.log(
            JSON.stringify(result) === JSON.stringify(output) ? 'SUCCESS ' : 'ERROR ',
            'input', JSON.stringify(input),
            'output', JSON.stringify(output),
            'result', JSON.stringify(result),
        );
    });
};

test();